packages/core/src/lib/core/mutations/access-control.ts

```diff
@@ -64,240 +63,224 @@ export async function checkUniqueItemExists(
 }
 
-export async function getAccessControlledItemForDelete(
-  list: InitialisedList,
+async function enforceListLevelAccessControl ({
+  context,
+  operation,
+  list,
+  item,
+  inputData,
+}: {
   context: KeystoneContext,
-  uniqueWhere: UniquePrismaFilter,
-  accessFilters: boolean | InputFilter
-) {
-  const operation = 'delete' as const;
-  // Apply the filter access control. Will throw an accessDeniedError if the item isn't found.
-  const item = await getFilteredItem(list, context, uniqueWhere!, accessFilters, operation);
-
-  // Apply item level access control
-  const access = list.access.item[operation];
-  const args = { operation, session: context.session, listKey: list.listKey, context, item };
-
+  operation: 'create' | 'update' | 'delete'
+  list: InitialisedList,
+  item: BaseItem | undefined,
+  inputData: Record<string, unknown>
+}) {
   // List level 'item' access control
-  let result;
+  let listResult: unknown; // should be boolean, but dont trust, it might accidentally be a filter
   try {
-    result = await access(args);
+    if (operation === 'create') {
+      const listAccessControl = list.access.item[operation];
+      listResult = await listAccessControl({
+        operation,
+        session: context.session,
+        listKey: list.listKey,
+        context,
+        inputData,
+      });
+    } else if (operation === 'update' && item !== undefined) {
+      const listAccessControl = list.access.item[operation];
+      listResult = await listAccessControl({
+        operation,
+        session: context.session,
+        listKey: list.listKey,
+        context,
+        item,
+        inputData,
+      });
+    } else if (operation === 'delete' && item !== undefined) {
+      const listAccessControl = list.access.item[operation];
+      listResult = await listAccessControl({
+        operation,
+        session: context.session,
+        listKey: list.listKey,
+        context,
+        item,
+      });
+    }
   } catch (error: any) {
     throw extensionError('Access control', [
-      { error, tag: `${args.listKey}.access.item.${args.operation}` },
+      { error, tag: `${list.listKey}.access.item.${operation}` },
     ]);
   }
 
-  const resultType = typeof result;
+  // short circuit the safe path
+  if (listResult === true) return;
 
-  // It's important that we don't cast objects to truthy values, as there's a strong chance that the user
-  // has accidentally tried to return a filter.
-  if (resultType !== 'boolean') {
+  if (typeof listResult !== 'boolean') {
     throw accessReturnError([
       {
-        tag: `${args.listKey}.access.item.${args.operation}`,
-        returned: resultType,
+        tag: `${list.listKey}.access.item.${operation}`,
+        returned: typeof listResult,
       },
     ]);
   }
 
-  if (!result) {
-    throw accessDeniedError(
-      `You cannot perform the '${operation}' operation on the item '${JSON.stringify(
-        uniqueWhere
-      )}'. It may not exist.`
```

Comment:
I think this was the wrong error, we know the `item` exists - if we're trying to make that fact opaque, we can call `missingItem` to throw and make that explicit